File: src/webhook/circuit_breaker.py
import time
from dataclasses import dataclass
from typing import Dict, List

import structlog

logger = structlog.get_logger()


@dataclass
class CircuitState:
    failures: int = 0
    opened_at: float = 0.0
    state: str = "closed"  # closed, open, half_open
    last_success: float = 0.0


class CircuitBreaker:
    def __init__(self):
        self._states: Dict[str, CircuitState] = {}

    def _get_state(self, url: str) -> CircuitState:
        if url not in self._states:
            self._states[url] = CircuitState()
        return self._states[url]

    def allow(self, url: str, recovery_secs: int) -> bool:
        st = self._get_state(url)
        if st.state == "closed":
            return True
        elif st.state == "open":
            if time.time() - st.opened_at > recovery_secs:
                st.state = "half_open"
                logger.info("Circuit transitioning to half-open for probe", url=url)
                return True
            return False
        elif st.state == "half_open":
            # Only allow one probe
            return False
        return True

    def record_success(self, url: str):
        st = self._get_state(url)
        if st.state != "closed":
            logger.info("Circuit closed (recovery successful)", url=url)
        st.failures = 0
        st.state = "closed"
        st.last_success = time.time()

    def record_failure(self, url: str, threshold: int):
        st = self._get_state(url)
        if st.state == "half_open":
            st.state = "open"
            st.opened_at = time.time()
            logger.warning("Circuit re-opened (probe failed)", url=url)
        else:
            st.failures += 1
            if st.failures >= threshold and st.state != "open":
                st.state = "open"
                st.opened_at = time.time()
                logger.error("Circuit opened", url=url, failures=st.failures)

    def get_state(self, url: str) -> dict:
        st = self._get_state(url)
        return {
            "state": st.state,
            "failures": st.failures,
            "opened_at": st.opened_at,
            "last_success": st.last_success,
        }

    def all_urls(self) -> List[str]:
        return list(self._states.keys())

    def reset(self, url: str):
        st = self._get_state(url)
        st.failures = 0
        st.state = "closed"
        st.opened_at = 0.0
        logger.info("Circuit manually reset", url=url)


_breaker = CircuitBreaker()


def get_circuit_breaker() -> CircuitBreaker:
    return _breaker

File: src/webhook/circuit_breaker.rs
use dashmap::DashMap;
use std::sync::Arc;
use std::sync::OnceLock;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Clone, PartialEq, Eq)]
pub enum State {
    Closed,
    Open,
    HalfOpen,
}

impl State {
    pub fn as_str(&self) -> &'static str {
        match self {
            State::Closed => "closed",
            State::Open => "open",
            State::HalfOpen => "half_open",
        }
    }
}

#[derive(Clone)]
pub struct CircuitState {
    pub failures: u32,
    pub opened_at: f64,
    pub state: State,
    pub last_success: f64,
}

impl Default for CircuitState {
    fn default() -> Self {
        Self {
            failures: 0,
            opened_at: 0.0,
            state: State::Closed,
            last_success: 0.0,
        }
    }
}

fn current_time() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
}

/// A high-performance, thread-safe Circuit Breaker implemented in Rust.
/// Uses DashMap for lock-free concurrency across all Python threads.
#[derive(Clone)]
pub struct CircuitBreaker {
    states: Arc<DashMap<String, CircuitState>>,
}

impl CircuitBreaker {
    pub fn new() -> Self {
        CircuitBreaker {
            states: Arc::new(DashMap::new()),
        }
    }

    pub fn allow(&self, url: &str, recovery_secs: f64) -> bool {
        let mut st = self
            .states
            .entry(url.to_string())
            .or_insert_with(CircuitState::default);

        match st.state {
            State::Closed => true,
            State::Open => {
                if current_time() - st.opened_at > recovery_secs {
                    st.state = State::HalfOpen;
                    true
                } else {
                    false
                }
            }
            State::HalfOpen => false,
        }
    }

    pub fn record_success(&self, url: &str) {
        let mut st = self
            .states
            .entry(url.to_string())
            .or_insert_with(CircuitState::default);
        st.failures = 0;
        st.state = State::Closed;
        st.last_success = current_time();
    }

    pub fn record_failure(&self, url: &str, threshold: u32) {
        let mut st = self
            .states
            .entry(url.to_string())
            .or_insert_with(CircuitState::default);

        if st.state == State::HalfOpen {
            st.state = State::Open;
            st.opened_at = current_time();
        } else {
            st.failures += 1;
            if st.failures >= threshold && st.state != State::Open {
                st.state = State::Open;
                st.opened_at = current_time();
            }
        }
    }

    pub fn get_state(&self, url: &str) -> Result<serde_json::Value, Box<dyn std::error::Error>> {
        let st = self
            .states
            .entry(url.to_string())
            .or_insert_with(CircuitState::default);

        Ok(serde_json::json!({
            "state": st.state.as_str(),
            "failures": st.failures,
            "opened_at": st.opened_at,
            "last_success": st.last_success,
        }))
    }

    pub fn all_urls(&self) -> Vec<String> {
        self.states.iter().map(|kv| kv.key().clone()).collect()
    }

    pub fn reset(&self, url: &str) {
        let mut st = self
            .states
            .entry(url.to_string())
            .or_insert_with(CircuitState::default);
        st.failures = 0;
        st.state = State::Closed;
        st.opened_at = 0.0;
    }
}

static BREAKER: OnceLock<CircuitBreaker> = OnceLock::new();

pub fn get_circuit_breaker() -> CircuitBreaker {
    BREAKER.get_or_init(|| CircuitBreaker::new()).clone()
}
